File: src/main/java/rwt/fview/FileData.java
package rwt.fview;

import java.io.File;
import java.io.RandomAccessFile;

class CachedBytes {
   public final long which;
   public final byte[] bytes;

   public CachedBytes(long w, byte[] b) { which = w; bytes = b; }
}

class CircularCache {
    private final CachedBytes[] cache;
    private int idx;

    CircularCache(int size) {
         idx = 0;
         cache = new CachedBytes[size];
         for(int idx = 0; idx < cache.length; ++idx) {
               cache[idx] = new CachedBytes(-1L, null);
         }
    }


    byte[] lookup(long which) {
       byte[] answer = null;
       for(int idx = 0; idx < cache.length; ++idx) {
           if(cache[idx].which == which) {
		answer = cache[idx].bytes;
                break;
           }
       } 
       return answer;
    }

    void insert(long which, byte[] bytes) {
        cache[idx] = new CachedBytes(which, bytes);
        idx++;
        if(idx == cache.length) {
            idx = 0;
        }
    }

    void clear() {
        for(int idx = 0; idx < cache.length; ++idx) {
            cache[idx] = new CachedBytes(-1, null);
        }
    }
}

/**
 * Created by richa on 10/11/2016.
 */
class FileData {

    private final RandomAccessFile src;
    private final CircularCache cache;
    private int chunkSize;

    FileData(File file, int cSize) throws java.io.IOException {
         src = new RandomAccessFile(file, "r");
         cache = new CircularCache(5);
         chunkSize = cSize;
    }


    void changeChunkSize(int newSz) {
        chunkSize = newSz;
        cache.clear();
    }

    private byte[] readChunk(long which) throws java.io.IOException {
        byte[] ans = new byte[chunkSize];
        src.seek( chunkSize * which );
        src.readFully(ans);
        return ans;
    }

    byte[] fetchData(long which) throws java.io.IOException {
         byte[] answer = cache.lookup(which);
         if(answer == null) {
              answer = readChunk(which);
              cache.insert(which, answer);
         }
         return answer;
    }

    void close() {
        try {
           src.close();
        } catch(Exception e) {  e.printStackTrace(); }
    }
}

File: src/rwt/fview/FileData.kt
package rwt.fview

import java.io.File
import java.io.RandomAccessFile

private class CachedBytes(val which : Long, val bytes : ByteArray)

private class CircularCache(sz : Int) {
    private val cache = Array<CachedBytes>(sz, { CachedBytes(-1, ByteArray(0))})
    private var idx = 0

    fun lookup(which : Long) = cache.find { it.which == which }?.bytes

    fun insert(which : Long, bytes : ByteArray) {
        cache[idx] = CachedBytes(which, bytes)
        idx++
        if(idx == cache.size) {
            idx = 0
        }
    }

    fun clear() {
        for(i in cache.indices) {
            cache[i] = CachedBytes(-1, ByteArray(0))
        }
    }
}

/**
 * Created by richa on 10/11/2016.
 */
internal class FileData(file : File, private var chunkSize : Int) {
    private val src = RandomAccessFile(file, "r")
    private val cache = CircularCache(5)

    fun changeChunkSize(newSz : Int) {
        chunkSize = newSz
        cache.clear()
    }

    private fun readChunk(which : Long) : ByteArray {
        val ans = ByteArray(chunkSize)
        src.seek( chunkSize.toLong() * which )
        src.readFully(ans)
        return ans
    }

    fun fetchData(which: Long) : ByteArray {
         return cache.lookup(which) ?: readChunk(which).apply { cache.insert(which, this) }
    }

    fun close() {
        src.close()
    }
}